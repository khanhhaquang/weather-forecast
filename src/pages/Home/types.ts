export type WeatherData = {
  id: number;
  min_temp: number;
  max_temp: number;
  applicable_date: Date;
  weather_state_name: string;
  weather_state_abbr: string;
};

export type CityDataResponse = {
  woeid: number;
};

export type WeatherDataResponse = {
  consolidated_weather: WeatherData[];
};
