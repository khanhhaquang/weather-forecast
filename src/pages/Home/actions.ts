export const getWeather = (city: string) => ({
  type: 'weather/request',
  city,
});
