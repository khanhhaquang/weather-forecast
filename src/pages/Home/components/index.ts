export { default as WeatherCard } from './WeatherCard';
export { default as SearchInput } from './SearchInput';
export { default as Error } from './Error';
export { default as LoadingSection } from './LoadingSection';
export { default as WeatherPanel } from './WeatherPanel';
