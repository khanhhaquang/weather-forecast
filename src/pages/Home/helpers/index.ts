// temporarily support yyyy/mm/dd
export const getWeekDay = (date: Date): string => {
  const weekDaysArr = [
    'Sunday',
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday',
  ];
  return weekDaysArr[date.getDay()];
};

export const roundVal = (value: number): number => Math.round(value);
