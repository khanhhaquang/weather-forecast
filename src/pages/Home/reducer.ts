import { WeatherData } from './types';

type THome = {
  isLoading: boolean;
  data: WeatherData[] | null;
  error: any;
};

const initialState: THome = {
  isLoading: false,
  data: null,
  error: null,
};

const reducer = (state = initialState, action: { type: string; data?: any }) => {
  switch (action.type) {
    case 'weather/request':
      return {
        ...state,
        isLoading: true,
        data: null,
        error: null,
      };
    case 'weather/success':
      return {
        ...state,
        isLoading: false,
        data: action.data,
        error: null,
      };
    case 'weather/error':
      return {
        ...state,
        isLoading: false,
        data: null,
        error: action.data,
      };
    default:
      return state;
  }
};

export const getErrorMessage = (state: any): string => state.home.error?.message || '';
export const getWeatherData = (state: any): WeatherData[] | [] => state.home.data || [];
export const getIsLoading = (state: any): boolean => state.home.isLoading;

export default reducer;
