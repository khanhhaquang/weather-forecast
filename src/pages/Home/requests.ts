import { weatherApi } from 'api';
import { AxiosResponse } from 'axios';

export const apiRequestLocation = (city: string): Promise<AxiosResponse> =>
  weatherApi.get(`/location/search/?query=${city}`);

export const apiRequestWeather = (woeid: number): Promise<AxiosResponse> =>
  weatherApi.get(`/location/${woeid}/`);
