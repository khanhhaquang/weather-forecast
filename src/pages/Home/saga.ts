import { put, takeLatest } from 'redux-saga/effects';
import { weatherApi } from 'api';
import axios from 'axios';

function* fetchWeatherSaga(action: { type: string; city: string }) {
  try {
    const cancelSource = axios.CancelToken.source;
    const data = yield call(
      weatherApi.get,
      `/location/search/?query=${action.city}`,
      {
        cancelToken: cancelSource.token,
      },
    );
    yield put({ type: 'weather/success', data });
  } catch (error) {
    yield put({ type: 'weather/error', data: error });
  }
}

function* watch() {
  yield takeLatest('weather/fetch', fetchWeatherSaga);
}

export default watch;
