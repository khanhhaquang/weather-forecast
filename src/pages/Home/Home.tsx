import React from 'react';
import styled from 'styled-components';
import { SearchInput, Error, LoadingSection, WeatherPanel } from './components';

const StyledWrapper = styled.div`
  max-width: 600px;
  margin: auto;
`;

const WeatherWrapper = styled.div`
  width: 100%;
  margin-top: 16px;
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  column-gap: 8px;
`;

const Home: React.FC = () => (
  <StyledWrapper>
    <SearchInput />
    <Error />
    <WeatherWrapper>
      <LoadingSection />
      <WeatherPanel />
    </WeatherWrapper>
  </StyledWrapper>
);

export default Home;
