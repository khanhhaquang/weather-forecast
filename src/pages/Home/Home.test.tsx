import { render, screen } from 'utils/test';
import Home from './Home';

describe('Home', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Home />, {});
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('search-input')).toBeInTheDocument();
  });
});
