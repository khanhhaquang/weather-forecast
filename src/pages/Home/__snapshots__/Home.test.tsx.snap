// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-jSFjdj EJfNd"
  >
    <input
      class="sc-hKFxyN dkcVOX"
      data-testid="search-input"
      placeholder="Which city do you want to check ?"
      value=""
    />
    <div
      class="sc-gKAaRy jYknsn"
    />
  </div>
</DocumentFragment>
`;
